tests/bmtestspolar/symfixedrankpolarfactory.m

```diff
@@ -0,0 +1,215 @@
+%       Year         = {2011},
+%       Organization = {{ICML}}
+%     }
+% For second-order geometry, please cite the Manopt paper as well as the research paper:
+%     @Article{mishra2014fixedrank,
+%       Title   = {Fixed-rank matrix factorizations and {Riemannian} low-rank optimization},
+%       Author  = {Mishra, B. and Meyer, G. and Bonnabel, B. and Sepulchre, R.},
+%       Journal = {Computational Statistics},
+%       Year    = {2014},
+%       Number  = {3-4},
+%       Pages   = {591--621},
+%       Volume  = {29},
+%       Doi     = {10.1007/s00180-013-0464-z}
+%     }
+%
+%
+% See also symfixedrankYYfactory sympositivedefinitefactory
+    
+    % This file is part of Manopt: www.manopt.org.
+    % Original author: Bamdev Mishra, April 05, 2019.
+    % Contributors:
+    % Change log:
+    %
+    
+    M.name = @() sprintf('UBU'' quotient manifold of %dx%d matrices of rank %d', m, m, k);
+    
+    M.dim = @() m*k - k*(k-1)/2;
+    
+    % Choice of the metric on the orthnormal space is motivated by the symmetry present in the
+    % space. The metric on the positive definite space is its natural metric.
+    M.inner = @(X, eta, zeta) eta.U(:).'*zeta.U(:)  ...
+        + trace( (X.B\eta.B) * (X.B\zeta.B) );
+    
+    M.norm = @(X, eta) sqrt(M.inner(X, eta, eta));
+    
+    M.dist = @(x, y) error('symfixedrankpolarfactory.dist not implemented yet.');
+    
+    M.typicaldist = @() 10*k;
+    
+    skew = @(X) .5*(X-X');
+    symm = @(X) .5*(X+X');
+    stiefel_proj = @(U, H) H - U*symm(U'*H);
+    
+    M.egrad2rgrad = @egrad2rgrad;
+    function rgrad = egrad2rgrad(X, egrad)
+        rgrad.U = stiefel_proj(X.U, egrad.U);
+        rgrad.B = X.B*symm(egrad.B)*X.B;
+    end
+    
+    
+    M.ehess2rhess = @ehess2rhess;
+    function Hess = ehess2rhess(X, egrad, ehess, eta)
+        
+        % Riemannian gradient for the factor B.
+        rgrad.B = X.B*symm(egrad.B)*X.B;
+        
+        % Directional derivatives of the Riemannian gradient.
+        Hess.U = ehess.U - eta.U*symm(X.U'*egrad.U);
+        Hess.U = stiefel_proj(X.U, Hess.U);
+        
+        Hess.B = X.B*symm(ehess.B)*X.B +  2*symm(eta.B*symm(egrad.B)*X.B);
+        
+        % Correction factor for the non-constant metric on the factor B.
+        Hess.B = Hess.B - symm(eta.B*(X.B\rgrad.B));
+        
+        % Projection onto the horizontal space.
+        Hess = M.proj(X, Hess);
+    end
+    
+    
+    M.proj = @projection;
+    function etaproj = projection(X, eta)
+        % First, projection onto the tangent space of the total space.
+        eta.U = stiefel_proj(X.U, eta.U);
+        eta.B = symm(eta.B);
+        
+        % Then, projection onto the horizontal space.
+        SS = X.B*X.B;
+        AS = X.B*(skew(X.U'*eta.U) - 2*skew(X.B\eta.B))*X.B;
+        
+        % Compute skew-symmetric Omega.
+        % To solve the system SS*Omega + Omega*SS - B*Omega*B = AS
+        Omega = mylinearsystem(SS, AS);
```

Comment:
If we apply inv(B) ... inv(B) to this equation, then inv(B)*AS*inv(B) simplifies to just it's middle part (skew symmetric) (let's call it Z here), and we have

B Omega inv(B) + inv(B) * Omega B - Omega = Z

Should we be solving this equation in this form rather than the current one? We can do the same trick of computing [V, D] = eig(B), and substituting VDV' in the equation, to get a simple form, where 

S = V' Omega V

and S(i, j) = Z(i, j) / (D(i, i)/D(j, j) + D(j, j)/D(i, i) - 1)

which can surely be computed without for-loops.

In any case, I don't know if it would be much better, but it does seem to save some computations, in particular the squaring of B (which may be bad for conditioning) and hitting AS with B on both sides.

Reply:
will try out. basically, I wanted to avoid inv(B) in the equation and hence squared the terms out. In the meantime, I am marking this resolved.